### PR TITLE
Fix some problems with watches that return a v1.Status

### LINKED ISSUE #516: Watch fails to parse V1Status because of incorrect successful parsing of watch type
We're working with 3.0.0 and will try reproducing with 4.0.0; although, the relevant code in Watch hasn't changed.

I have a watch for config maps and because we keep the watches running for a long time, we sometimes get 410 Gone exceptions because the resourceVersion is "too old".  Unfortunately, the V1Status with that information is getting swallowed because the JSON deserializaton incorrectly succeeds.

For instance, we captured the following inbound response:

Padded plaintext after DECRYPTION:  len = 186
0000: 62 34 0D 0A 7B 22 74 79   70 65 22 3A 22 45 52 52  b4..."type":"ERR
0010: 4F 52 22 2C 22 6F 62 6A   65 63 74 22 3A 7B 22 6B  OR","object":."k
0020: 69 6E 64 22 3A 22 53 74   61 74 75 73 22 2C 22 61  ind":"Status","a
0030: 70 69 56 65 72 73 69 6F   6E 22 3A 22 76 31 22 2C  piVersion":"v1",
0040: 22 6D 65 74 61 64 61 74   61 22 3A 7B 7D 2C 22 73  "metadata":..,"s
0050: 74 61 74 75 73 22 3A 22   46 61 69 6C 75 72 65 22  tatus":"Failure"
0060: 2C 22 6D 65 73 73 61 67   65 22 3A 22 74 6F 6F 20  ,"message":"too 
0070: 6F 6C 64 20 72 65 73 6F   75 72 63 65 20 76 65 72  old resource ver
0080: 73 69 6F 6E 3A 20 34 35   30 31 34 39 39 38 20 28  sion: 45014998 (
0090: 34 35 30 31 36 30 31 37   29 22 2C 22 72 65 61 73  45016017)","reas
00A0: 6F 6E 22 3A 22 47 6F 6E   65 22 2C 22 63 6F 64 65  on":"Gone","code
00B0: 22 3A 34 31 30 7D 7D 0A   0D 0A                    ":410.....

However, the code from Watch here:

      try {
        return json.deserialize(line, watchType);
      } catch (JsonParseException ex) {
        Type statusType = new TypeToken<Response<V1Status>>() {}.getType();
        Response<V1Status> status = json.deserialize(line, statusType);
        return new Response<T>(status.type, status.object);
      }

... succeeds in deserializing that input as V1ConfigMap (in our case):

type: ERROR, status: <none>, object: class V1ConfigMap {\n    apiVersion: v1\n    binaryData: null\n    data: null\n    kind: Status\n    metadata: class V1ObjectMeta {\n        annotations: null\n        clusterName: null\n        creationTimestamp: null\n        deletionGracePeriodSeconds: null\n        deletionTimestamp: null\n        finalizers: null\n        generateName: null\n        generation: null\n        initializers: null\n        labels: null\n        name: null\n        namespace: null\n        ownerReferences: null\n        resourceVersion: null\n        selfLink: null\n        uid: null\n    }\n}

### PR DESCRIPTION
**Pull request by brendandburns**
Fixes #516 

@rjeberhard @kondapally1989

### CODE REVIEW

File: util/src/test/java/io/kubernetes/client/util/WatchTest.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (rjeberhard) [inline util/src/test/java/io/kubernetes/client/util/WatchTest.java]: Copyright year 2019

Comment (brendandburns) [reply util/src/test/java/io/kubernetes/client/util/WatchTest.java]: fixed.

**Comment by kondapally1989** [comment]
@brendandburns 
/lgtm.

will it be better to have watchevent(https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#watchevent-v1-meta) and decide to return object based on field type?

File: util/src/main/java/io/kubernetes/client/util/Watch.java [hunk 2]
```diff
@@ -125,18 +126,60 @@ private Watch(JSON json, ResponseBody body, Type watchType, Call call) {
...
-      }
+      return parseLine(line);
     } catch (IOException e) {
       throw new RuntimeException("IO Exception during next method.", e);
     }
   }
 
+  protected boolean isStatus(String line) throws IOException {
+    boolean found = false;
+    JsonReader reader = new JsonReader(new StringReader(line));
```
Comment (yue9944882) [inline util/src/main/java/io/kubernetes/client/util/Watch.java]: this doubles the overhead of deserializing normal watch event 👀

Comment (brendandburns) [reply util/src/main/java/io/kubernetes/client/util/Watch.java]: No, this is a streaming parser, so it's more like string processing. The real cost of parsing is in the object allocation and reflection. (at least that's my belief...)

File: util/src/test/java/io/kubernetes/client/util/WatchTest.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    status.add("metadata", metadata);
+    status.addProperty("kind", "Status");
+    status.addProperty("apiVersion", "v1");
+    status.addProperty("status", "failure");
+    status.addProperty("message", "too old resource version");
+    status.addProperty("reason", "Gone");
+    status.addProperty("code", 410);
+
+    JsonObject obj = new JsonObject();
+    obj.addProperty("type", "ERR");
```
[the remark this answers is not in the document]
Comment (brendandburns) [reply util/src/test/java/io/kubernetes/client/util/WatchTest.java]: done.

**Comment by brendandburns** [comment]
@kondapally1989 yeah, we should probably migrate to `WatchEvent` but not in this PR.

**Comment by brendandburns** [comment]
Comments addressed, please re-check.

Thanks!
--brendan

**Comment by yue9944882** [comment]
/lgtm

ps: labeling `tide/squash` will make the pull automatically squashed before getting into master

**Pull request merged**